File: pset6/vigenere/vigenere.py
# Ruchella Kock
# 12460796
# This program will take an alphabetical key which will cipher a text the user fills in

import sys
from cs50 import get_string


def cipher(each_character, each_key_character, y):
    """
    enciphers each character of a string
    """
    alphabetical_cipher = ((ord(each_key_character.lower()) - 97) + (ord(each_character) - y)) % 26
    cipher_character = chr(alphabetical_cipher + y)
    print(f"{cipher_character}", end="")


def main():
    # program will not run if it doesnt get 2 command line arguments
    if len(sys.argv) != 2:
        sys.exit("Usage: vigenere.py key")

    # check if there is an integer in the key
    key = sys.argv[1]
    if not key.isalpha():
        sys.exit("Usage: vigenere.py alphabetical key")

    # prompt user for text
    plain_text = get_string("plaintext: ")
    print("ciphertext: ", end="")

    A, a, j = 65, 97, 0
    length_key = len(key)
    # go over every character of the string and execute cipher accordingly
    for character in plain_text:
        each_key_character = key[j % length_key]
        if character.isupper():
            cipher(character, each_key_character, A)
            j = j + 1
        elif character.islower():
            cipher(character, each_key_character, a)
            j = j + 1
        else:
            print(f"{character}", end="")
    print()


if __name__ == "__main__":
    main()

File: pset6/vigenere/vigenere.c
/*
Ruchella Kock
12460796

This programs takes an alphabetical key in the command line argument 1, it uses this key to encipher a word the user fills in

*/

#include <cs50.h>
#include <ctype.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#define A 65
#define a 97
// function to encipher each character of a string
int cipher(char each_character, char each_key_character, int x, int y)
{
    int alphabetical_value_key = (int) each_key_character - x;
    int alphabetical_value = (int) each_character - y;
    int alphabetical_cipher = (alphabetical_value + alphabetical_value_key) % 26;
    int ASCII = (alphabetical_cipher + y);
    printf("%c", ASCII);
    return ASCII;
}

int main(int argc, string argv[])
{
    // program will stop if 1 or more than 2 arguments are filled in command line
    if (argc == 1 || argc > 2)
    {
        printf("Error\n");
        return 1;
    }

    // program will not run if there is an integer in the key
    string key = argv[1];
    for (int i = 0, num = strlen(key); i < num; i++)
    {
        if (isdigit(key[i]))
        {
            printf("Error\n");
            return 1;
        }
    }

    // prompt user for a string to encipher
    string plain_text = get_string("plaintext: ");
    printf("ciphertext: ");

    // iterate over each character of the string the user filled in
    for (int i = 0, j = 0, n = strlen(plain_text); i < n; i++)
    {
        char each_character = (plain_text[i]);
        // if the character of the string is upper case then determine if key is upper case or lower case
        if (isupper(each_character))
        {
            // depending on whether key is upper or lower case the parameters of function cipher changes
            char each_key_character = (key[j % strlen(key)]);
            if (isupper(each_key_character))
            {
                cipher(each_character, each_key_character, A, A);
                j++;
            }
            if (islower(each_key_character))
            {
                cipher(each_character, each_key_character, a, A);
                j++;
            }
        }
        // if the character of the string is lower case then determine if key is upper case or lower case
        else if (islower(each_character))
        {
            char each_key_character = (key[j % strlen(key)]);
            // depending on whether key is upper or lower case the parameters of function cipher changes
            if (islower(each_key_character))
            {
                cipher(each_character, each_key_character, a, a);
                j++;
            }
            else if (isupper(each_key_character))
            {
                cipher(each_character, each_key_character, A, a);
                j++;
            }
        }
        // if the character is not an alphabetical value then print that character
        else if (isalpha(each_character) != 1)
        {
            printf("%c", each_character);
        }
    }
    printf("\n");
}